File: internal/db/events.go
package db

import (
	"database/sql"
	"time"
)

// EventFilter filters event listings.
type EventFilter struct {
	Month         string
	ApplicationID int64
	EventType     string
	Start         *time.Time
	End           *time.Time
}

// EventWithApplication includes the application display fields for an event.
type EventWithApplication struct {
	Event
	CompanyName  string
	PositionName string
}

// CreateEvent inserts a new schedule event.
func (db *Database) CreateEvent(event *Event) error {
	res, err := db.conn.Exec(
		`INSERT INTO events (application_id, event_type, round, scheduled_at, duration, location, notes) VALUES (?, ?, ?, ?, ?, ?, ?)`,
		event.ApplicationID, event.EventType, event.Round, nullableTime(event.ScheduledAt), event.Duration, event.Location, event.Notes,
	)
	if err != nil {
		return err
	}
	event.ID, _ = res.LastInsertId()
	event.CreatedAt = time.Now()
	return nil
}

// GetEvent retrieves a single event by ID.
func (db *Database) GetEvent(id int64) (*Event, error) {
	var event Event
	var scheduledAt sql.NullTime
	err := db.conn.QueryRow(
		`SELECT id, application_id, event_type, round, scheduled_at, duration, location, notes, created_at FROM events WHERE id = ?`, id,
	).Scan(&event.ID, &event.ApplicationID, &event.EventType, &event.Round, &scheduledAt, &event.Duration, &event.Location, &event.Notes, &event.CreatedAt)
	if err != nil {
		return nil, err
	}
	event.ScheduledAt = timePtrFromNull(scheduledAt)
	return &event, nil
}

// ListEvents lists events joined with their application display fields.
func (db *Database) ListEvents(filter EventFilter) ([]EventWithApplication, error) {
	query := `SELECT e.id, e.application_id, e.event_type, e.round, e.scheduled_at, e.duration, e.location, e.notes, e.created_at, a.company_name, a.position_name
		FROM events e
		JOIN applications a ON a.id = e.application_id`
	var args []interface{}
	var where []string

	if filter.Month != "" {
		start, err := time.Parse("2006-01", filter.Month)
		if err != nil {
			return nil, err
		}
		end := start.AddDate(0, 1, 0)
		where = append(where, "e.scheduled_at >= ? AND e.scheduled_at < ?")
		args = append(args, start, end)
	}
	if filter.ApplicationID > 0 {
		where = append(where, "e.application_id = ?")
		args = append(args, filter.ApplicationID)
	}
	if filter.EventType != "" {
		where = append(where, "e.event_type = ?")
		args = append(args, filter.EventType)
	}
	if filter.Start != nil {
		where = append(where, "e.scheduled_at >= ?")
		args = append(args, *filter.Start)
	}
	if filter.End != nil {
		where = append(where, "e.scheduled_at < ?")
		args = append(args, *filter.End)
	}
	for i, clause := range where {
		if i == 0 {
			query += " WHERE "
		} else {
			query += " AND "
		}
		query += clause
	}
	query += ` ORDER BY e.scheduled_at ASC, e.id ASC`

	rows, err := db.conn.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []EventWithApplication
	for rows.Next() {
		var event EventWithApplication
		var scheduledAt sql.NullTime
		if err := rows.Scan(&event.ID, &event.ApplicationID, &event.EventType, &event.Round, &scheduledAt, &event.Duration, &event.Location, &event.Notes, &event.CreatedAt, &event.CompanyName, &event.PositionName); err != nil {
			return nil, err
		}
		event.ScheduledAt = timePtrFromNull(scheduledAt)
		out = append(out, event)
	}
	return out, rows.Err()
}

// ListEventsByApplication lists all events for a single application.
func (db *Database) ListEventsByApplication(applicationID int64) ([]EventWithApplication, error) {
	return db.ListEvents(EventFilter{ApplicationID: applicationID})
}

// UpdateEvent updates an existing schedule event.
func (db *Database) UpdateEvent(event *Event) error {
	res, err := db.conn.Exec(
		`UPDATE events SET application_id = ?, event_type = ?, round = ?, scheduled_at = ?, duration = ?, location = ?, notes = ? WHERE id = ?`,
		event.ApplicationID, event.EventType, event.Round, nullableTime(event.ScheduledAt), event.Duration, event.Location, event.Notes, event.ID,
	)
	if err != nil {
		return err
	}
	return errNoRowsWhenUnchanged(res)
}

// DeleteEvent removes an event by ID.
func (db *Database) DeleteEvent(id int64) error {
	res, err := db.conn.Exec(`DELETE FROM events WHERE id = ?`, id)
	if err != nil {
		return err
	}
	return errNoRowsWhenUnchanged(res)
}

func nullableTime(v *time.Time) interface{} {
	if v == nil {
		return nil
	}
	return *v
}

func timePtrFromNull(v sql.NullTime) *time.Time {
	if !v.Valid {
		return nil
	}
	t := v.Time
	return &t
}

func errNoRowsWhenUnchanged(res sql.Result) error {
	affected, err := res.RowsAffected()
	if err != nil {
		return err
	}
	if affected == 0 {
		return sql.ErrNoRows
	}
	return nil
}

File: src/offerpilot/repositories/events.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Optional

from sqlalchemy import select
from sqlalchemy.orm import Session, sessionmaker

from offerpilot.models import Application, Event


@dataclass
class EventCreate:
    application_id: int
    event_type: str
    scheduled_at: datetime
    duration_minutes: int
    round: int = 0
    location: str = ""
    notes: str = ""


@dataclass
class EventWithApplication:
    event: Event
    company_name: str
    position_name: str


class EventsRepository:
    def __init__(self, session_factory: sessionmaker[Session]):
        self._session_factory = session_factory

    def create(self, data: EventCreate) -> Event:
        event = Event(
            application_id=data.application_id,
            event_type=data.event_type,
            round=data.round,
            scheduled_at=data.scheduled_at,
            duration=_duration_string(data.duration_minutes),
            location=data.location,
            notes=data.notes,
        )
        with self._session_factory() as session:
            session.add(event)
            session.commit()
            session.refresh(event)
            return event

    def list(
        self,
        month: str = "",
        application_id: int = 0,
        event_type: str = "",
    ) -> list[EventWithApplication]:
        statement = (
            select(Event, Application.company_name, Application.position_name)
            .join(Application, Application.id == Event.application_id)
            .order_by(Event.scheduled_at.asc(), Event.id.asc())
        )
        if month:
            statement = statement.where(Event.scheduled_at >= f"{month}-01")
        if application_id > 0:
            statement = statement.where(Event.application_id == application_id)
        if event_type:
            statement = statement.where(Event.event_type == event_type)

        with self._session_factory() as session:
            rows = session.execute(statement).all()
            return [
                EventWithApplication(event=row[0], company_name=row[1], position_name=row[2])
                for row in rows
            ]

    def get(self, event_id: int) -> Optional[Event]:
        with self._session_factory() as session:
            return session.get(Event, event_id)

    def update(self, event_id: int, data: EventCreate) -> Optional[Event]:
        with self._session_factory() as session:
            event = session.get(Event, event_id)
            if event is None:
                return None
            event.application_id = data.application_id
            event.event_type = data.event_type
            event.round = data.round
            event.scheduled_at = data.scheduled_at
            event.duration = _duration_string(data.duration_minutes)
            event.location = data.location
            event.notes = data.notes
            session.commit()
            session.refresh(event)
            return event

    def delete(self, event_id: int) -> bool:
        with self._session_factory() as session:
            event = session.get(Event, event_id)
            if event is None:
                return False
            session.delete(event)
            session.commit()
            return True


def _duration_string(minutes: int) -> str:
    return f"{minutes}m"


def duration_minutes(duration: str) -> int:
    return int(duration.removesuffix("m") or "0")

